src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,227 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using Microsoft.AspNetCore.Sockets;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    public class JsonHubProtocol : IHubProtocol
+    {
+        private const string ResultPropertyName = "result";
+        private const string InvocationIdPropertyName = "invocationId";
+        private const string TypePropertyName = "type";
+        private const string ErrorPropertyName = "error";
+        private const string TargetPropertyName = "target";
+        private const string NonBlockingPropertyName = "nonBlocking";
+        private const string ArgumentsPropertyName = "arguments";
+
+        private const int InvocationMessageType = 1;
+        private const int ResultMessageType = 2;
+        private const int CompletionMessageType = 3;
+
+        // ONLY to be used for application payloads (args, return values, etc.)
+        private JsonSerializer _payloadSerializer;
+
+        public MessageType MessageType => MessageType.Text;
+
+        /// <summary>
+        /// Creates an instance of the <see cref="JsonHubProtocol"/> using the specified <see cref="JsonSerializer"/>
+        /// to serialize application payloads (arguments, results, etc.). The serialization of the outer protocol can
+        /// NOT be changed using this serializer.
+        /// </summary>
+        /// <param name="payloadSerializer">The <see cref="JsonSerializer"/> to use to serialize application payloads (arguments, results, etc.).</param>
+        public JsonHubProtocol(JsonSerializer payloadSerializer)
+        {
+            _payloadSerializer = payloadSerializer;
+        }
+
+        public bool TryParseMessage(ReadOnlySpan<byte> input, IInvocationBinder binder, out HubMessage message)
+        {
+            // TODO: Need a span-native JSON parser!
+            using (var memoryStream = new MemoryStream(input.ToArray()))
+            {
+                message = ParseMessage(memoryStream, binder);
+            }
+            return true;
```

Comment:
Actually, this doesn't need to be a try method at all, it's like our parser/formatter methods but unlike there we know exactly how much data to parse, so there's no need return `false` to indicate more data is required.